src/main/java/org/wildfly/security/http/impl/FormAuthenticationMechanism.java

```diff
@@ -162,7 +162,7 @@ private void attemptAuthentication(HttpServerRequest request) throws HttpAuthent
         String username = request.getFirstParameterValue(USERNAME);
         String password = request.getFirstParameterValue(PASSWORD);
 
-        if (username == null || password == null) {
+        if (username == null || username.trim().length() == 0 || password == null) {
```

Comment:
Is the trim() necessary?  If so should we update our reference to username to the trimmed value?

Reply:
I think that trim is necessary to catch user names containing only spaces, but don't know if we want to trim valid user names.

Reply:
What happens if a whitespace only username is received?

Reply:
I think that we need to throw username is missing exception.

Reply:
But what does happen if it is empty?  Do we already handle that cleanly?  The validation here seems slightly above the original symptom being addressed,

Reply:
You are right that the original issue is that we need to check if the username is empty and not when whitespace(s) only username is received, because javax.security.auth.callback.NameCallback throws exception only in this case[1]. 
If you think that we are doing an extra validation by throwing the same error for whitespace(s) only usernames, I'll remove it. If we remove it, do we need to handle this case with another error or do nothing because this is a valid value. 

[1] http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/8u40-b25/javax/security/auth/callback/NameCallback.java/#89

Reply:
That is really what I am trying to ask as you are already testing this, I would hope we have sufficient validation in other places that would pick up an empty username - the error being fixed here is more due to null getting passed into a Callback but that same callback would accept empty string.

Reply:
I believe whitespace-only username is valid and should be accepted - at least on authentication mechanisms side.